### ISSUE TITLE
HTML DOM Representation

### ISSUE THREAD

**Issue by DavePearce** [Whiley/Whiley2JavaScript#42]
_(@altmattr --- this may also be of interest to you)_

My initial efforts to represent the DOM in Whiley have provided some interesting observations.  Firstly, we need some way to interact natively with the DOM (see #41).  Secondly, the DOM has an inbuilt concept of a union type.  For example, here is an abridged version of the `Node` interface:

```
public type Node is &{
    int nodeType,
    js_string nodeName,
    (null | Element) parent,
    Node[] childNodes,
    ...
}
```

From this, we can see two interesting things: `js_string` and `null | Element`.

Both of these pose interesting questions.  At this stage, I've just modelled a `js_string` using an `int` because this give me a sensible (if not slightly broken) translation.  Likewise, at this stage, the union type `null | Element` is modelled correctly because we are not using tagged unions.  However, as soon as we employ tagged unions (which is my current plan) this will break.

**Comment by DavePearce** [comment]
One possible solution here is to support a concept of an _untagged_ or _raw_ union.  This is something I've thought about before.  The syntax would be: `(null || Element)`.  In principle, this should only happen between record types which have a common initial sequence.  But, that doesn't work for `(null || Element)` anyways. Hmmmmmmmmmm

**Comment by DavePearce** [comment]
The other option, of course, is just to accept that we have to have a shim around the DOM to enable interaction from Whiley.  That could work, but would add some overhead.  Potentially in many cases this could be circumvented.

**Comment by altmattr** [comment]
Yep, I would suggest that approach.  If you are taking the elm approach then you are in whiley-land the whole time until you inspect the dom to update it - so the cost won't be too bad.

matt

________________________________

**Comment by DavePearce** [comment]
Well, yes, but then in Elm land you are stuck to a specific model.  Whereas if we can represent the actual DOM somehow, then we can interact with it and choose our own models.

Still, even having a shim around the DOM ... I'm not sure how that would work.  I can do it using the virtual DOM approach of Elm, but I don't know how to enable more direct access.

**Comment by altmattr** [comment]
I'll continue my experiments and keep bugging you when I find problems then 🙂

matt

________________________________

**Comment by DavePearce** [comment]
Ok, maybe I could figure this out using a `native` type.  Maybe I should have a go and see if I can do that.  Currently fighting with various minor problems in the compiler in an effort to get an actual single-page app working :(

**Comment by DavePearce** [comment]
Also, I thought you were against the Elm approach??

**Comment by altmattr** [comment]
My attitude is that models like this are wonderful, and far too restrictive for real web developers.  If you want to build an impressive system, then the elm approach will serve you well.  If you are pitching to real web developers then you will find them building more and more elaborate ways to get around the fact that you have not given them direct access to the DOM.

I don't mean the _real_ DOM though, the whiley-land equivalent of the DOM is fine.

So, it is my dual-personality that is showing.  Functional programmer by day (who loves an elegant model), writer of real-world applications by night (who owns more spaghetti-code than he cares to admit)🙂
________________________________
